### PR TITLE
Support transforming inherited methods

### LINKED ISSUE #14: Consider supporting prototype methods
Any chance you would be willing to consider supporting `pify`-ing the prototypally-inherited methods of passed objects?

The main issue that I'm trying to solve is that code such as:

``` javascript
var lovebug = pify(new Car("Herbie"));
```

works if the `Car` constructor sets/copies/binds methods to `this`, but not if the methods are inherited via `prototype`.  So, unless I've missed something, `pify` users must either depend on the implementation of `Car` or roll their own method iteration wrapper for `pify` (and/or `pify` the prototype, then `pify` any methods set by the constructor).

If `pify`-ing inherited methods sounds agreeable but you'd need a PR to consider, I can work one up.  I would envision the methods being copied from the prototype to the new target object if a boolean option is set, but alternatives would be fine with me.

Thanks for considering,
Kevin

### PR DESCRIPTION
**Pull request by kevinoid**
This makes it possible to transform objects which are instances of a class without knowing which methods are present on the instance object and which are inherited through the prototype chain, as discussed in #14.  Include/Exclude is handled the same as for non-inherited methods.

In addition to transforming inherited methods, when the `inherited` option is enabled the result object is created with the prototype of the original object so that it behaves the same as the original object for `instanceof` checks.  I think this fits with the intended use of `inherited`, but if you would prefer a separate option or a separate PR, let me know and I'll split it out.

This PR also copies all properties from the original object to the target, even if excluded, to match the current behavior.  It may make more sense to avoid copying values and/or excluded methods which would be inherited, both to avoid unnecessary copies and so any updates to the prototype are shared.  But I left that behavior out of this PR for simplicity.  Let me know if you would prefer that behavior.

Thanks for considering,
Kevin

Fixes #14

### CODE REVIEW

**Comment by SamVerschueren** [comment]
Also did a PR (https://github.com/sindresorhus/pify/pull/17) but closed it in favour of this one. Should've looked better.

But like to discuss the implementation. As you can see in my PR, I simply changed `Object.keys()` to `for...in`. The question is, why isn't that enough? Is it really necessary to implement the flag? Just wondering and looking for feedback on that one :).

File: index.js [hunk 1]
```diff
@@ -48,15 +48,32 @@ var pify = module.exports = function (obj, P, opts) {
 		return opts.include ? opts.include.some(match) : !opts.exclude.some(match);
 	};
 
-	var ret = typeof obj === 'function' ? function () {
-		if (opts.excludeMain) {
-			return obj.apply(this, arguments);
-		}
+	var ret;
```
Comment (SamVerschueren) [inline index.js]:
```
var ret = {};
```

And remove last `else`.

File: index.js [hunk 2]
```diff
@@ -48,15 +48,32 @@ var pify = module.exports = function (obj, P, opts) {
...
-	} : {};
+	var props;
+	if (opts.inherited) {
+		props = [];
+		for (var prop in obj) {	// eslint-disable-line guard-for-in
+			props.push(prop);
+		}
+	} else {
+		props = Object.keys(obj);
+	}
```
Comment (SamVerschueren) [inline index.js]: Not sure about this block. If `inherited` is set to `true`, it iterates twice over the properties. Once in the `for...in`, and once in the `reduce` function below.

You could extract the function and do it like this

```js
if(opts.inherited) {
     for (var prop in obj) {
         handler(ret, prop);
     }

     return ret;
}

return Object.keys(obj).reduce(handler, ret);
```

> NOTE: Just a quick sample, might contain issues :)

But before changing this, we should discuss the `inherited` property first.

**Comment by kevinoid** [comment]
I think the main motivation of adding this behavior as an option is to avoid breaking existing code which is not expecting it and avoid an unnecessary cost for users who don't need it.

Personally, I'd change the default to `true` in the next major version release, since I think it's a better default, but that's definitely debatable and up to the project maintainers.

File: index.js (hunk 1 above)
Comment (kevinoid) [inline index.js]: I disagree with the often-unnecessary object instantiation, but not strongly.  If the project maintainers prefer your style, I'll update it.

File: index.js (hunk 2 above)
Comment (kevinoid) [inline index.js]: Sure, that would work.  Alternatively, `reduce` could be avoided altogether by using that imperative style in both cases (or moving the `if` inside the loop to combine the cases).  It would probably perform better on microbenchmarks.

The commit was intended to minimize changes by keeping the reduce loop as-is, but I have no opinion on which style is preferable.  If the project maintainers would prefer an imperative style, I'll update it.

File: index.js (hunk 1 above)
Comment (sindresorhus) [inline index.js]: > var ret = {};
And remove last else.

:+1:

Comment (kevinoid) [inline index.js]: Roger wilco.

**Comment by sindresorhus** [comment]
@SamVerschueren Just thought about this and looked at my original comment in the issue. I think it should be an option and `false` by default, like it currently is, since I would never use this option personally.

> Personally, I'd change the default to true in the next major version release, since I think it's a better default, but that's definitely debatable and up to the project maintainers.

Doubtful that will happen as I personally would never use this option and I hope there's not going to be another major release.

File: index.js (hunk 2 above)
Comment (sindresorhus) [inline index.js]: > Sure, that would work. Alternatively, reduce could be avoided altogether by using that imperative style in both cases (or moving the if inside the loop to combine the cases). It would probably perform better on microbenchmarks.

:+1: I don't care about the microperf improvements, but I think it would make the code more readable.

**Comment by kevinoid** [comment]
@sindresorhus @SamVerschueren Changes implemented.  What do you think of this version?

**Comment by schnittstabil** [comment]
Do we need resp. should copy non-enumerables? E.g. 

```js
function C(value) {
  Object.defineProperty(this, 'value', {value});
}

const instance = new C(42);
const pInstance = fn(instance, {inherited: true});

instance.value // => 42
pInstance.value // => undefined
```

**Comment by kevinoid** [comment]
> Do we need resp. should copy non-enumerables?

I'd suggest a `nonenumerable` option which would be independent of `inherited`.  I can add it to this PR if you'd like.

Would you expect the non-enumerable properties of the original object to be non-enumerable on the target object?  If preserving the own enumerable properties is a concern, note that this PR currently copies inherited properties rather than `pify`-ing the prototype chain separately.  So the own enumerable properties of the result are a superset of the the own enumerable properties of the original object.  I can change this if it's an issue.

**Comment by sindresorhus** [comment]
> Do we need resp. should copy non-enumerables? E.g.

No. Not interested in adding an option either. Already have more options than I'd like.

**Comment by SamVerschueren** [comment]
> Just thought about this and looked at my original comment in the issue. I think it should be an option and `false` by default, like it currently is, since I would never use this option personally.

Just wondering what the downside is of not supporting the prototypes by default (thus without an option). 

Take for example the [AWS sdk](https://aws.amazon.com/sdk-for-node-js/) (which I need this for).

This is how it looks like to upload something to S3 without `pify`.

```js
var AWS = require('aws-sdk');

var s3 = new AWS.S3();

s3.createBucket({Bucket: 'myBucket'}, function () {
	var params = {Bucket: 'myBucket', Key: 'myKey', Body: 'Hello!'};

	s3.putObject(params, function (err) {
		if (err) {
			console.log(err);
			return;
		}

		console.log('Successfully uploaded data to myBucket/myKey');
	});
});
```

The implementation of `S3` might look like this

```js
function S3() {
	this.config = {};
	// some other properties
}

S3.prototype.createBucket = function (params, cb) {

};

S3.prototype.putObject = function (params, cb) {

};
```

If I instantiate a new object `new S3()`, then `createBucket` and `putObject` are functions of the object and thus imo should be pified when using `pify(new S3())`.

I don't really see the downside of not pifying them by default. If someone wants to pify `fs`, nothing extra will get promisified because it doesn't have prototype functions.

Also thinking in terms of number of options. That's why I'm asking myself, is it really necessary to put this behaviour in an option?

**Comment by schnittstabil** [comment]
@kevinoid I don't know, I've just thought it might be related. Sigh, ES2015 Proxies will solve all that copy-crap one day.

**Comment by sindresorhus** [comment]
@SamVerschueren Yes, would be better without an option, but I guess that would be a breaking change? And I really don't want to do another major now. So how about we add it as an option now, and just default it to true always in the next major.

It should also not copy these props: https://github.com/petkaantonov/bluebird/blob/3915a2bfa4188baacb43345d5b084983b9dfbccc/src/promisify.js#L13-L22

**Comment by SamVerschueren** [comment]
Not sure if it's a breaking change. Can't think of a use case where `pify` works at the moment, and with adding this it breaks. It just pifies extra methods, right? Or maybe I'm missing something :p.

**Comment by sindresorhus** [comment]
@SamVerschueren I'm just commenting from:

> @kevinoid: I think the main motivation of adding this behavior as an option is to **avoid breaking existing code which is not expecting it** and avoid an unnecessary cost for users who don't need it.

I personally can't think of how it would break anything, but might be something I'm missing.

**Comment by SamVerschueren** [comment]
@sindresorhus Alright, an option for now will do then. We can always remove it if a new major is being released in the future.

**Comment by kevinoid** [comment]
> So how about we add it as an option now, and just default it to true always in the next major.

Great!

> It should also not copy these props: https://github.com/petkaantonov/bluebird/blob/3915a2bfa4188baacb43345d5b084983b9dfbccc/src/promisify.js#L13-L22

Should those properties be excluded unconditionally or only when `obj` is a function?  It looks like bluebird excludes them when `promisify` is called but not when `promisifyAll` is called.

> I personally can't think of how it would break anything, but might be something I'm missing.

The breakage I was thinking about would be if code called `pify` on an instance but was still using the un-promisified inherited methods on the result.  It shouldn't be a common case, but I tend to be conservative about backwards-compatibility.  Feel free to consider or ignore it as you like.

**Comment by sindresorhus** [comment]
> Should those properties be excluded unconditionally or only when obj is a function? It looks like bluebird excludes them when promisify is called but not when promisifyAll is called.

Actually, never mind. Doesn't look like we need to since we don't copy the descriptors. https://github.com/petkaantonov/bluebird/commit/ce4d91f29658538455aa9180132307e8860a3e76

> The breakage I was thinking about would be if code called pify on an instance but was still using the un-promisified inherited methods on the result. It shouldn't be a common case, but I tend to be conservative about backwards-compatibility. Feel free to consider or ignore it as you like.

Yes, that's a good example.

**Comment by kevinoid** [comment]
> Actually, never mind. Doesn't look like we need to since we don't copy the descriptors. petkaantonov/bluebird@ce4d91f

That's true.  Good point.

**Comment by sindresorhus** [comment]
@kevinoid ping :)

**Comment by kevinoid** [comment]
@sindresorhus Is there something waiting on me?  If so, my apologies, I thought I had gotten everything.

**Comment by SamVerschueren** [comment]
I think it should be an option for now instead of by default.

**Comment by kevinoid** [comment]
> I think it should be an option for now instead of by default.

The PR requires passing `options.inherited` to have any effect. I don't think it changes anything by default.

**Comment by SamVerschueren** [comment]
Ah, totally missed that one! Not sure would should be done then...

**Comment by akupila** [comment]
Is there a reason the pifyfied functions are not bound to the the original instance? This creates a problem when the pifyfied class tries to access its own instance through `this.xxx`. I ran into this problem with the AWS SDK.

With the current implementation this happens:
```js
const s3 = pify(new AWS.S3(), { inherited: true })

s3.listBuckets()
.then((results) => {
  console.error('result')
  console.log(results)
})
.catch((err) => {
  console.error('error')
  console.error(err)
})

// error
// [TypeError: SignerClass is not a function]
// (happens because getSignerClass gets wrapped to return a promise)
```
```js
const s3 = pify(new AWS.S3(), { inherited: true, exclude: ['getSignerClass'] })

// result
// {}
// ???
```

A quick solution for this is to set the scope in `processFn` to the original object.

I did some testing and this works:
```js
var scope = this;
if (typeof obj === 'function') {
	ret = function () {
		if (opts.excludeMain) {
			return obj.apply(this, arguments);
		}

		return processFn(obj, P, opts).apply(this, arguments);
	};
} else if (opts.inherited) {
	ret = Object.create(Object.getPrototypeOf(obj));
	scope = obj;
}

for (var key in obj) {
	if (opts.inherited || obj.hasOwnProperty(key)) {
		var x = obj[key];
		ret[key] = typeof x === 'function' && filter(key) ? processFn(x, P, opts).bind(scope) : x;
	}
}
``` 
Note the scope being set with `opts.inherited` and `processFn.bind(scope)`.

Interestingly enough `getSignerClass` doesn't need to be excluded with this either as it still lives in the non-pifyfied instance. Not sure if this is a bug or a feature :)

I guess one downside can be the additional overhead of `bind` if it's not needed; could only `bind()` if `opts.inherited`. Are there other problems with this? Another solution would be to pass the scope as an argument to `processFn`.

For completeness with these changes:
```js
const s3 = pify(new AWS.S3(), { inherited: true })

s3.listBuckets()
.then((results) => {
  console.error('result')
  console.log(results)
})
.catch((err) => {
  console.error('error')
  console.error(err)
})

// result
// { Buckets:
//   [ { Name: '< bucket name >,
//       CreationDate: Fri Mar 04 2016 12:11:02 GMT+0100 (CET) },
//     { Name: '< bucket name >',
//       CreationDate: Tue Mar 08 2016 12:38:27 GMT+0100 (CET) },
//     { Name: '< bucket name >',
//       CreationDate: Fri Mar 04 2016 18:24:52 GMT+0100 (CET) },
//     { Name: '< bucket name >',
//       CreationDate: Wed Mar 02 2016 11:03:02 GMT+0100 (CET) } ],
//  Owner:
//   { DisplayName: 'akupila',
//     ID: '< id >' } }
```

**Comment by akupila** [comment]
Should add: Not sure _why_ S3 is returning nothing when `getSignerClass` is excluded. I guess there's something internal to AWS SDK that gets broken in pify but no errors show up and the aws code is a bit hard to follow..

Another way of approaching this would be to whitelist the methods to pify:
```js
const s3 = pify(new AWS.S3(), { inherited: true, include: ['listBuckets'] })

s3.listBuckets()
.then((results) => {
  // ok!
})
```

This works but it's a bit annoying to need to remember to do this for each method.

Sorry if i'm hijacking this PR with an AWS SDK specific issue! I'd imagine this will apply to other (bigger) libraries too though

**Comment by sdgluck** [comment]
I feel like the `inherited` option is a bit ambiguous and also that adding this functionality makes the module less focused, so I forked this project and created [`pify-proto`](https://github.com/sdgluck/pify-proto) which is `pify` that deals exclusively with promisifying the prototype of an object.

@akupila `pify-proto` handles binding methods to their original context so should handle your use-case just fine.

Cheers all. (And thanks for a great module, Sindre + contributors.)

**Comment by Turbo87** [comment]
@kevinoid @SamVerschueren what is the status of this PR? anyone still working on it? seems to me like there is not much missing to push it over the finish line.

**Comment by kevinoid** [comment]
@Turbo87 I'm not aware of anything else which needs to be done.  I think it's just waiting on a decision from @sindresorhus.  Feel free to weigh in on it.

**Comment by Turbo87** [comment]
@kevinoid I was under the impression that the comment above by @akupila (https://github.com/sindresorhus/pify/pull/16#issuecomment-203310874) still needed to be fixed.

**Comment by kevinoid** [comment]
@Turbo87 If @sindresorhus wants bind behavior (conditionally or unconditionally) as part of this PR, I'm happy to add it.  But I didn't want to complicate the issue by adding it without discussion first.

**Comment by schnittstabil** [comment]
Maybe [bind-methods](https://github.com/sindresorhus/bind-methods) or a similar library solve this too? For example:

```js
// instead of:
const s3 = pify(new AWS.S3(), {inherited: true});

// we do 
const s3 = pify(bindMethods(new AWS.S3()));

// or with a thinkable library:
const s3 = pify(bindAllMethodsIncludingPrototype(new AWS.S3()));

// or with a bind-methods option:
const s3 = pify(bindMethods(new AWS.S3(), {inherited: true}));
```

Please note how [`bindMethods`](https://github.com/sindresorhus/bind-methods/blob/44c174ea04cbc02d85aacdd1714b2811caa4007f/index.js#L5-L11) works:

```js
for (const key of Object.keys(obj)) {
	const val = obj[key];

	if (typeof val === 'function') {
		obj[key] = val.bind(context);
	}
}
```

**Comment by schnittstabil** [comment]
As far as I can see, [`bind-all`](https://github.com/segmentio/bind-all/blob/master/lib/index.js#L7-L13) provides the desired behavior:

```js
const bindAll = require('bind-all');
const s3 = pify(bindAll(new AWS.S3()));
```

At least it solves issues with [`gm`](https://github.com/aheckmann/gm) (mentioned at #36):

```js
const gm = require("gm").subClass({ imageMagick: true });
const bindAll = require('bind-all');

const gmP = (...args) => pify(bindAll(gm(...args)));

test("Resize JPEG with cjpeg", async t => {
    //…
    const out = await gmP(destPath).size();
    t.is(out.width, 200);
});
```

**Comment by kevinoid** [comment]
@schnittstabil Good find.  That would work, as would any library which flattens the inheritance hierarchy.  [`bind-all`](https://github.com/segmentio/bind-all) obviously has the additional effect that methods will be called with `this` set to the original object rather than the promisified one, but that's probably what you would want in most cases anyway.

If pify won't support inheritance and bind, I think that documenting the behavior and recommending [`bind-all`](https://github.com/segmentio/bind-all) would be the next best solution.

**Comment by sindresorhus** [comment]
Way overdue, sorry, but finally merged. @kevinoid I decided to go with your solution, but by default without the option. I'm going to do a major release very soon.

@kevinoid @schnittstabil @SamVerschueren @kevva Could you take a look at https://github.com/sindresorhus/pify/compare/v2.3.0...master at let me know if you have any comment before I do a new major release? 🌈

**Comment by kevinoid** [comment]
No worries.  Better late than never.  Looks good to me.  Thanks @sindresorhus!

**Comment by schnittstabil** [comment]
:+1: `master` LGTM.

**Comment by SamVerschueren** [comment]
Sorry, late to the party :). Looks good though!
